core/src/test/java/org/apache/spark/launcher/SparkLauncherSuite.java

```diff
@@ -183,6 +183,28 @@ public void testChildProcLauncher() throws Exception {
     assertEquals(0, app.waitFor());
   }
 
+  @Test
+  public void testThreadLauncher() throws Exception {
+    // This test is failed on Windows due to the failure of initiating executors
+    // by the path length limitation. See SPARK-18718.
+    assumeTrue(!Utils.isWindows());
+
+    launcher
+      .setMaster("local")
+      .setAppResource(SparkLauncher.NO_RESOURCE)
+      .setConf(SparkLauncher.DRIVER_EXTRA_JAVA_OPTIONS,
+        "-Dfoo=bar -Dtest.appender=childproc")
+      .setConf(SparkLauncher.DRIVER_EXTRA_CLASSPATH, System.getProperty("java.class.path"))
+      .setMainClass(SparkLauncherTestApp.class.getName())
+      .launchAsThread(true)
+      .addAppArgs("proc");
+    final Process app = launcher.launch();
```

Comment:
What is this testing? `launch()` will always launch a child process. Which indicates two problems:

- this test is not testing anything that hasn't been tested before.
- `SparkLauncher` should probably be throwing an error if you use `.launchAsThread(true)` and then call `launch()`.